Add restricted method text (#1217)
MetaMask/metamask-docs: wallet/concepts/wallet-api.md

request users' Ethereum accounts, read data from blockchains the user is connected to, suggest
that the user sign messages and transactions, and more.

:::note
MetaMask supports [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963), which introduces an
alternative discovery mechanism to the `window.ethereum` injected provider.
This alternative mechanism enables dapps to support [wallet interoperability](../wallet-interoperability)
by discovering multiple injected wallet providers in a user's browser.

> We recommend using this mechanism for provider discovery.

Reviewer: Please revert these changes, which revert merged changes in #1214.

:::

The MetaMask Ethereum provider API contains the following:

- [Properties](../reference/provider-api.md#properties) - The provider contains a property that
  detects if a user has MetaMask installed.
- [Methods](../reference/provider-api.md#methods) - The provider contains methods that dapps can call.

>   The [`window.ethereum.request(args)`](../reference/provider-api.md#windowethereumrequestargs)

Reviewer: Please revert this change, which reverts a merged change in #1214.

## JSON-RPC API

> MetaMask uses the [`window.ethereum.request(args)`](../reference/provider-api.md#windowethereumrequestargs)

Reviewer: Please revert this change, which reverts a merged change in #1214.

:::note
All RPC method requests can return errors.
Make sure to handle errors for every call to

> [`window.ethereum.request(args)`](../reference/provider-api.md#windowethereumrequestargs).

Reviewer: Please revert this change, which reverts a merged change in #1214.

The following methods are restricted:

- [eth_accounts](/wallet/reference/eth_accounts) - Gaining permission requires calling `wallet_requestPermissions`. 
Granting permissions for `eth_accounts` also provides permissions for the following methods:
  - [`eth_sendTransaction`](/wallet/reference/eth_sendTransaction)
  - [`personal_sign`](/wallet/reference/personal_sign)
  - [`eth_signTypedData_v4`](/wallet/reference/eth_signTypedData_v4)

:::info note
To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts).
This method automatically obtains permissions for `eth_accounts` through an internal `wallet_requestPermissions` call.
See [how to access a user's accounts](../how-to/connect/access-accounts.md) for more information.
:::

- [`wallet_snap`](/wallet/reference/wallet_snap) - Gaining permission requires calling `wallet_requestSnap`.
- [`wallet_invokeSnap`](/wallet/reference/wallet_invokeSnap) - Gaining permission requires calling `wallet_requestSnap`.
  
:::info note

> For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the example at the end of the [About the Snaps APIs](../../../../snaps/learn/about-snaps/apis.md#custom-json-rpc-apis) page.

Reviewer: ```suggestion
For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the example at the end of the [About the Snaps APIs](/snaps/learn/about-snaps/apis/#custom-json-rpc-apis) page.
```

Sorry for the confusion, but when linking between `wallet` and `snaps`, you need to use absolute links.

### Restricted methods

MetaMask introduced wallet permissions in [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255).
In this permissions system, each RPC method is restricted or unrestricted. 

Restricted methods are methods that cannot be called unless you have permission to do so using [`wallet_requestPermissions`](/wallet/reference/wallet_requestpermissions) or [wallet_requestSnaps](/wallet/reference/wallet_requestSnaps).

Under the hood, permissions are plain, JSON-compatible objects, with fields that are mostly used
internally by MetaMask.

> Outside of [Snaps restricted methods](/snaps/reference/rpc-api/#restricted-methods), there is the

Author: eth_accounts requires request permissions using ...
you need request permissions for the wallet_ snap / wallet_invokeSnap  using the wallet_request

restricted method [`eth_accounts`](/wallet/reference/eth_accounts), which allows you to access

> the user's Ethereum accounts. DApps must explicitly request user consent using `wallet_requestPermissions` before using `eth_accounts`.

Author: Dapp

Author: As well as add a NOTE about eth_requestAccounts

Author: @alexandratran suggestion 

The following methods are restricted:
- eth_accounts - Requires passing eth_accounts as a parameter to wallet_requestPermissions. This unlocks:
  - eth_sendTransaction
  - personal_sign
  - eth_signTypedData_v4
note: we recommend calling eth_requestAccounts, which internally calls wallet_requestPermissions on eth_accounts
- wallet_snap - Requires calling wallet_requestSnap
- wallet_invokeSnap - Requires calling wallet_requestSnap

To call wallet_snap or wallet_invokeSnap on a Snap, you need to first call wallet_requestSnap on that Snap
Example on the bottom of this section: https://docs.metamask.io/snaps/learn/about-snaps/apis/#custom-json-rpc-apis (edited)

Granting permission for `eth_accounts` also implicitly grants access to [`eth_sendTransaction`](/wallet/reference/eth_sendTransaction), [`personal_sign`](/wallet/reference/personal_sign), and [`eth_signTypedData`](/wallet/reference/eth_signTypedData). 

> 

Author: make sure no excessive line breaks

> ### Unrestricted methods

Author: remove the might still rely on to succeed since we are including this in the definition now

The following methods are restricted:

- [`eth_accounts`](/wallet/reference/eth_accounts) - Gaining permission requires calling `wallet_requestPermissions`. 
Granting permission for `eth_accounts` also grants permissions for the following methods:
  - [`eth_sendTransaction`](/wallet/reference/eth_sendTransaction)
  - [`personal_sign`](/wallet/reference/personal_sign)
  - [`eth_signTypedData_v4`](/wallet/reference/eth_signTypedData_v4)

:::caution important
To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts), which automatically asks for permission to use `eth_accounts` by calling `wallet_requestPermissions` internally.
See [how to access a user's accounts](../how-to/connect/access-accounts.md) for more information.
:::

- [`wallet_snap`](/wallet/reference/wallet_snap) - Gaining permission requires calling `wallet_requestSnap`.
- [`wallet_invokeSnap`](/wallet/reference/wallet_invokeSnap) - Gaining permission requires calling `wallet_requestSnap`.
  
:::info note

> For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the [requesting permissions from a dapp](/snaps/how-to/request-permissions/#request-permissions-from-a-dapp) example.

Reviewer: ```suggestion
For more information on using `wallet_snap` and `wallet_invokeSnap`, see the [how to request Snap permissions from a dapp](/snaps/how-to/request-permissions/#request-permissions-from-a-dapp).
```

[`request()`](../reference/provider-api.md#request).
:::

> Dapps communicate with MetaMask through JSON-RPC methods. 
> These methods are divided into the following:
> 
> - [Restricted methods](#restricted-methods) -  Require user consent for actions that impact assets or data (for example initiating a transaction).
> - [Unrestricted methods](#unrestricted-methods) - Allow dapps to perform basic actions without permission (for example retrieving a public address).

Reviewer: ```suggestion
The RPC methods are divided into the following:

- [Restricted methods](#restricted-methods) -  Require user consent for actions that impact assets or data (for example, initiating a transaction).
- [Unrestricted methods](#unrestricted-methods) - Allow dapps to perform basic actions without permission (for example, retrieving a public address).
```

### Restricted methods

MetaMask implements permissions based on [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255) to enhance security for when users interact with dapps. 
This requires that dapps obtain user consent before accessing certain features. 
Under the hood, permissions are plain, JSON-compatible objects, with fields that are mostly used
internally by MetaMask.

Restricted methods are methods that cannot be called unless you have permission to do so using [`wallet_requestPermissions`](/wallet/reference/wallet_requestpermissions) or [`wallet_requestSnaps`](/wallet/reference/wallet_requestSnaps).

The following methods are restricted:

> - [eth_accounts](/wallet/reference/eth_accounts) - Gaining permission requires calling `wallet_requestPermissions`. 
> Granting permissions for `eth_accounts` also provides permissions for the following methods:

Reviewer: ```suggestion
- [`eth_accounts`](/wallet/reference/eth_accounts) - Gaining permission requires calling `wallet_requestPermissions`. 
Granting permission for `eth_accounts` also grants permissions for the following methods:
```

  - [`eth_sendTransaction`](/wallet/reference/eth_sendTransaction)
  - [`personal_sign`](/wallet/reference/personal_sign)
  - [`eth_signTypedData_v4`](/wallet/reference/eth_signTypedData_v4)

> :::info note
> To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts).
> This method automatically obtains permissions for `eth_accounts` through an internal `wallet_requestPermissions` call.

Reviewer: ```suggestion
:::caution important
To access accounts, we recommend using [`eth_requestAccounts`](/wallet/reference/eth_requestAccounts), which automatically asks for permission to use `eth_accounts` by calling `wallet_requestPermissions` internally.
```

See [how to access a user's accounts](../how-to/connect/access-accounts.md) for more information.
:::

- [`wallet_snap`](/wallet/reference/wallet_snap) - Gaining permission requires calling `wallet_requestSnap`.
- [`wallet_invokeSnap`](/wallet/reference/wallet_invokeSnap) - Gaining permission requires calling `wallet_requestSnap`.
  
:::info note

> For more information on requesting permission to connect to `wallet_snap` and `wallet_invokeSnap`, see the example at the end of the [About the Snaps APIs](/snaps/learn/about-snaps/apis/#custom-json-rpc-apis) page.

Reviewer: Maybe copy the example referenced here to the [Request permissions from a dapp](https://docs.metamask.io/snaps/how-to/request-permissions/#request-permissions-from-a-dapp) Snaps section and link there instead.

Author: The entire section or just related to the getting the permissions? Am I duplicating the content for the custom JSON-RPC APIs?

[`request()`](../reference/provider-api.md#request).
:::

Dapps communicate with MetaMask through JSON-RPC methods. 
These methods are divided into the following:

- **Unrestricted methods** - Allow dapps to perform basic actions without permission (for example retrieving a public address).
- **Restricted methods** - Require user consent for actions that impact assets or data (for example initiating a transaction).

### Restricted methods

MetaMask implements permissions based on [EIP-2255](https://eips.ethereum.org/EIPS/eip-2255) to enhance security for when users interact with dapps. 
This requires that dapps obtain user consent before accessing certain features. 
Under the hood, permissions are plain, JSON-compatible objects, with fields that are mostly used
internally by MetaMask.

> Restricted methods are methods that cannot be called unless you have permission to do so using [`wallet_requestPermissions`](/wallet/reference/wallet_requestpermissions) or [wallet_requestSnaps](/wallet/reference/wallet_requestSnaps).

Reviewer 2: Missed the \` \` on wallet_requestSnaps
